Add context menu and copy support for URL-based hyperlink text

Dimmitsaras opened the pull request:
- [x] Depends on https://github.com/ppy/osu/pull/24022

Addresses discussion https://github.com/ppy/osu/discussions/23328

This is probably not the best implementation but I did it anyways as an exercise. Second commit is just cleaning up with inspectcode.

Demo:

https://user-images.githubusercontent.com/35996393/235251801-77353f9b-8903-4c3f-a982-132e1efb5350.mp4

Dimmitsaras committed:
Added context menus to chat links

Dimmitsaras committed:
Added context menus to chat links

Joehuu commented on osu.Game/Online/Chat/DrawableLinkCompiler.cs:
```diff
@@ -52,6 +62,32 @@ private void load(OsuColour colours)
+                    items.Add(new OsuMenuItem("Open", MenuItemType.Highlighted, () => host.OpenUrlExternally(text)));
+                }
+
+                items.Add(new OsuMenuItem("Copy URL", MenuItemType.Standard, () =>
```
This doesn't work on wiki.

![osu!_OcsC28PbZo](https://user-images.githubusercontent.com/35318437/235284993-50ac64f5-84a3-40e0-b912-56875d1b42c1.gif)

Dimmitsaras replied:
Good catch, it seems like the many of the wiki links have different TooltipText values compared to the displayed hover tooltip. I specifically used TooltipText because it was an easy solution for the /np scenario, making the assumption that it should always correspond to the url. 
Is there a distinction between the two?

Dimmitsaras replied:
Im partway through an implementation that uses Link to derive LinkAction and Url to help with this case. I can also pass Link through LinkHandler to also make the highlighted action behave like a left click as noted below.

However it seems like the Wiki clickables are not Links. Is this an osu-wiki dependency? I might be stuck here.

Joehuu commented on osu.Game/Online/Chat/DrawableLinkCompiler.cs:
```diff
@@ -52,6 +62,32 @@ private void load(OsuColour colours)
+            get
+            {
+                List<MenuItem> items = new List<MenuItem>();
+                bool hasTooltip = TooltipText.ToString() != "";
```
Use `!string.IsNullOrEmpty()`.

Joehuu commented on osu.Game/Online/Chat/DrawableLinkCompiler.cs:
```diff
@@ -52,6 +62,32 @@ private void load(OsuColour colours)
+
+                if (!isChannelorLobbyLink)
+                {
+                    items.Add(new OsuMenuItem("Open", MenuItemType.Highlighted, () => host.OpenUrlExternally(text)));
```
I think the expected UX for highlighted menu items is that it behaves like left click. In this case, it should open the in-game overlays.

Joehuu commented on osu.Game/Online/Chat/DrawableLinkCompiler.cs:
```diff
@@ -52,6 +62,32 @@ private void load(OsuColour colours)
+
+                bool isChannelorLobbyLink = text.Contains("osu://chan") || text.Contains("osump://");
+
+                if (!isChannelorLobbyLink)
```
Doesn't need to exist if what I'm saying above is true.

frenzibyte replied:
Hmm, you make a good point. I'll try splitting it off the compiler.

Joehuu commented on osu.Game/Online/Chat/DrawableLinkCompiler.cs:
```diff
@@ -30,6 +34,12 @@ public partial class DrawableLinkCompiler : OsuHoverContainer
         private OverlayColourProvider overlayColourProvider { get; set; }
 
+        [Resolved]
+        private OnScreenDisplay onScreenDisplay { get; set; }
```
This is to fix the test error, but the `#nullable disable` line needs to be removed. See https://github.com/ppy/osu/wiki/Nullable-Reference-Types-(NRT).
```
        private OnScreenDisplay? onScreenDisplay { get; set; }
```

Joehuu reviewed (changes requested):
Not sure if this is the correct direction, but here's my review.

Dimmitsaras committed:
Wiki links broken, Highlighted option now corresponds to left click action

Dimmitsaras committed:
Functional wiki links

Dimmitsaras commented:
Thank you for your comments, they've been incredibly helpful!

I have implemented the default action to behave the same way as a left click. The text also adjusts between "Open, Visit and Join". This can be further expanded for specific LinkActions.

I tested it with a few wiki links across a few pages and it seems to function properly as well!

Dimmitsaras committed:
Wiki links now behave based off their LinkDetails

Joehuu commented on osu.Game/Online/Chat/DrawableLinkCompiler.cs:
```diff
@@ -52,6 +77,41 @@ private void load(OsuColour colours)
+                List<MenuItem> items = new List<MenuItem>();
+                string text;
+
+                switch (link.Action)
```
I would check all links as this is a very general class. I got this when right clicking the mapper name on a beatmap card:

![osu!_WMqHk6BKRM](https://user-images.githubusercontent.com/35318437/235334934-271416fd-68a2-4d50-a4cc-5f283d8d3937.png)

![rider64_gvcJqUptqR](https://user-images.githubusercontent.com/35318437/235334814-87cbfeb5-7d6d-4fd7-a828-27699d5ff2f5.png)

Joehuu commented on osu.Game/Online/Chat/DrawableLinkCompiler.cs:
```diff
@@ -1,44 +1,69 @@
 
+        private Link link { get; set; } = null!;
+
         [Resolved(CanBeNull = true)]
```
```suggestion
        [Resolved]
```

Joehuu commented on osu.Game/Online/Chat/DrawableLinkCompiler.cs:
```diff
@@ -1,44 +1,69 @@
+        [Resolved]
+        private GameHost host { get; set; } = null!;
+
+        [Resolved(CanBeNull = true)]
+        private ILinkHandler linkHandler { get; set; } = null!;
```
Should be:
```suggestion
        [Resolved]
        private ILinkHandler? linkHandler { get; set; }
```

Joehuu commented on osu.Game/Online/Chat/DrawableLinkCompiler.cs:
```diff
@@ -52,6 +77,41 @@ private void load(OsuColour colours)
+                        break;
+                }
+
+                items.Add(new OsuMenuItem(text, MenuItemType.Highlighted, () => linkHandler.HandleLink(link.Url)));
```
```suggestion
                items.Add(new OsuMenuItem(text, MenuItemType.Highlighted, () => linkHandler?.HandleLink(link.Url)));
```

Joehuu commented on osu.Game/Online/Chat/DrawableLinkCompiler.cs:
```diff
@@ -64,5 +124,13 @@ public LinkHoverSounds(HoverSampleSet sampleSet, List<Drawable> parts)
             public override bool ReceivePositionalInputAt(Vector2 screenSpacePos) => parts.Any(d => d.ReceivePositionalInputAt(screenSpacePos));
         }
+
+        public string GetUrlFromPart(List<Drawable> part)
```
Seems to be unused now.

Dimmitsaras committed:
Added null safeguard, fixed osu!supporter link

Dimmitsaras commented:
Thanks for your review!

You are right, it seems there are more cases of the DrawableLinkCompiler that might not provide links yet. I identified the user profile, changelog and beatmap pages as such.
This affects most links in those pages, from profile names to recent achievements or even the permalink, reply, report buttons.

For now i added a null check safeguard that defaults to those elements not being right clickable/not having a context menu.
This prevents catastrophic failure, but we also have to consider whether the permalink/reply/report buttons should actually be right clickable. While they are drawn as links, they are not traditional ones.

I also applied a fix to the "become an osu! supporter" link on the changelog, it's an isolated case so it should not regress anything.

Joehuu commented on osu.Game/Online/Chat/DrawableLinkCompiler.cs:
```diff
@@ -52,6 +77,45 @@ private void load(OsuColour colours)
+
+                    items.Add(new OsuMenuItem(text, MenuItemType.Highlighted, () => linkHandler?.HandleLink(link.Url)));
+
+                    items.Add(new OsuMenuItem("Copy URL", MenuItemType.Standard, () =>
```
I would make user links work (in beatmap cards, comments) or is it out of scope of this PR?

Joehuu commented:
Not sure if this PR is going to be delayed because it is using classes that should be reworked long ago, see https://github.com/ppy/osu/issues/11754 and referenced PR.

peppy commented on osu.Game/Graphics/Containers/LinkFlowContainer.cs:
```diff
@@ -123,7 +126,7 @@ protected override IEnumerable<Drawable> CreateDrawablesFor(TextFlowContainer te
         }
 
-        protected virtual DrawableLinkCompiler CreateLinkCompiler(ITextPart textPart) => new DrawableLinkCompiler(textPart);
+        protected virtual DrawableLinkCompiler CreateLinkCompiler(ITextPart textPart) => new DrawableLinkCompiler(textPart, currentLink);
```
This field usage is super super dodgy.

Instead, I think you should add a `LinkDetails` field in `TextLink` and pass it in in the constructor at:

https://github.com/ppy/osu/blob/7bc8908ca9c026fed1d831eb6e58df7624a8d614/osu.Game/Graphics/Containers/LinkFlowContainer.cs#L92

You can then consume it in the place you are currently.

frenzibyte replied:
Should be addressed in https://github.com/ppy/osu/pull/23334/commits/e92572b62792b92fef9803a85293000b82f5836a.

Dimmitsaras commented:
Thank you for your feedback,

I made a brief attempt in using LinkDetails instead, but it gets more intricate as LinkDetails.argument is detached from the URL in the case of internal links. While handleLink switches the LinkAction and can parse the argument accordingly, I see difficulty in getting the full URL for clipboard purposes
![image](https://user-images.githubusercontent.com/35996393/235432607-0fe34961-81fa-4b25-94d4-0df40f94baaf.png)

I agree with the sentiment that we should probably not increase work scope on classes pending rework.
That being said I would be happy to re-implement a similar solution once we get there, our interactions in this PR were very helpful

Joehuu commented:
If this PR doesn't get blocked because of `LinkDetails` further being used, I would make a reverse method of `MessageFormatter.GetLinkDetails(string url)` which is `MessageFormatter.GetUrl(LinkDetails linkDetails)` and properly go through every action/argument and use that for clipboard.

May need special handling of chat links (maybe using your current link implementation?) as they may be converted (e.g. old link `/u` -> new link `/users`) and may be bad UX.

frenzibyte committed:
Merge branch 'master' into hyperlink-right-click

frenzibyte committed:
Add "beatmapsets" to user dictionary

frenzibyte committed:
Refactor and centralise drawable link action

frenzibyte committed:
Add helper method to retrieve URL from `LinkDetails`

frenzibyte committed:
Only show context menu items if a URL is present

frenzibyte committed:
Add context menu containers to settings sections and beatmap info wedge

Beatmap info wedge for user profile links in "mapped by" text.
Settings sections for FAQ link in tablet settings.

frenzibyte commented on osu.Game/Online/Chat/DrawableLinkCompiler.cs:
```diff
@@ -1,57 +1,105 @@
         {
             IdleColour = overlayColourProvider?.Light2 ?? colours.Blue;
+
+            Action ??= () =>
+            {
+                Debug.Assert(Link.Action != LinkAction.Custom);
+
+                if (linkHandler != null)
+                    linkHandler.HandleLink(Link);
+                // fallback to handle cases where OsuGame is not available, ie. tournament client.
+                else if (Link.Action == LinkAction.External)
+                    host.OpenUrlExternally(Link.Argument.ToString());
+            };
```
This increases the scope of `DrawableLinkCompiler` a bit, but I think the name still stands? Maybe we could rename this to `DrawableLinkHandler`/`DrawableTextLinkHandler`, but I'm not sure how that covers the "effects" part of this component, so I left it as is.

bdach replied:
I'm going to have a rather large issue with all of this I'm afraid. Previously `DrawableLinkCompiler` was a simple wrapper for flowing text sprites to look like a link. Now it's resolving OSDs and handling clipboard and adding link-type-specific logic.

None of this should be here. It should be shipped off to a place specific to "URL-based hyperlinks".

frenzibyte replied:
Hmm, you make a good point. I'll try splitting it off the compiler.

frenzibyte commented on osu.Game/Online/Chat/DrawableLinkCompiler.cs:
```diff
@@ -1,57 +1,105 @@
+
+                return new MenuItem[]
+                {
+                    new OsuMenuItem("Open", MenuItemType.Highlighted, Action),
```
I've reverted the logic that decides the action text based on the destination of the link (https://github.com/ppy/osu/pull/23334/commits/e92572b62792b92fef9803a85293000b82f5836a#diff-fe2f5f4b0275915dcaadd5f4eedb0c524d17e906699d64362bdcd828de8c4055L90-L104), because it looked pretty much arbitrary.

bdach replied:
I'm going to have a rather large issue with all of this I'm afraid. Previously `DrawableLinkCompiler` was a simple wrapper for flowing text sprites to look like a link. Now it's resolving OSDs and handling clipboard and adding link-type-specific logic.

None of this should be here. It should be shipped off to a place specific for "URL-based hyperlinks".

Joehuu commented on osu.Game/Online/Chat/MessageFormatter.cs:
```diff
@@ -246,6 +246,44 @@ public static LinkDetails GetLinkDetails(string url)
+                    return $@"https://{WebsiteRootUrl}/beatmapsets/{link.Argument}";
+
+                case LinkAction.OpenUserProfile:
+                    return $@"https://{WebsiteRootUrl}/users/{link.Argument}";
```
This doesn't work on the beatmap wedge:
![osu!_jCKABii2Zg](https://github.com/ppy/osu/assets/35318437/df3802c1-4561-4fe2-ac01-23eaf7b45b9e)

Joehuu commented on osu.Game/Online/Chat/MessageFormatter.cs:
```diff
@@ -246,6 +246,44 @@ public static LinkDetails GetLinkDetails(string url)
+            switch (link.Action)
+            {
+                case LinkAction.External:
+                    return (string)link.Argument;
```
Should handle osu! links without root url:
![osu!_BE6muXsviO](https://github.com/ppy/osu/assets/35318437/ba03ce0b-054e-4998-9f85-a561be54b015)

frenzibyte committed:
Split context menu handling off `DrawableLinkCompiler`

frenzibyte committed:
Handle edge cases with certain links in game

frenzibyte committed:
Remove unused using directive

frenzibyte committed:
Mark OSD dependency as nullable

Joehuu commented on osu.Game/Online/Chat/MessageFormatter.cs:
```diff
@@ -246,6 +247,52 @@ public static LinkDetails GetLinkDetails(string url)
+                case LinkAction.OpenUserProfile:
+                    // todo: this is awkward. there should *only* be one type of argument per link action, but alas.
+                    return link.Argument is IUser user
+                        ? $@"https://{WebsiteRootUrl}/users/{user.Username}"
```
This should have the online id and fallback to username if it's not populated.

frenzibyte replied:
I'll much rather put the effort to remove `IUser` from being an argument of `OpenUserProfile` at this point.

frenzibyte committed:
Merge branch 'master' into hyperlink-right-click

frenzibyte committed:
Merge branch 'fix-open-user-profile-argument' into hyperlink-right-click

frenzibyte committed:
Revert handling for `IUser` arguments

frenzibyte committed:
Remove unused using directive

frenzibyte committed:
Fix failing test case

frenzibyte committed:
Merge branch 'fix-open-user-profile-argument' into hyperlink-right-click

frenzibyte committed:
Handle `IUser` arguments again

peppy committed:
Merge branch 'master' into hyperlink-right-click

peppy committed:
Add xmldoc for new `GetUrl` method

peppy commented on osu.Game/Overlays/Settings/SettingsSection.cs:
```diff
@@ -83,53 +84,58 @@ protected SettingsSection()
         private void load(OverlayColourProvider colourProvider)
         {
-            AddRangeInternal(new Drawable[]
+            AddInternal(new OsuContextMenuContainer
```
Shitting these all over the code base is pretty bad.

This one should be at a `SettingsOverlay` level if anything, IMHO.

Also it should have a comment mentioning what it's being used by or something. Or it should just be moved to global because this is ugly.

peppy commented on osu.Game/Screens/Select/BeatmapInfoWedge.cs:
```diff
@@ -199,116 +200,120 @@ private void load(LocalisationManager localisation)
                 const float top_height = 0.7f;
 
-                Children = new Drawable[]
+                Child = new OsuContextMenuContainer
```
Why is this required? Should be at song select or global.

peppy commented:
Code looks fine apart from these context containers.

Joehuu commented:
Been wanting to unblock this. I have a branch to address the issues when making `ContextMenuContainer` global:
- off-screen context menus when `IHasContextMenu` drawables are inside a scroll container (by hiding context menu on any scroll input
	- behavior is copying Windows Explorer and Edge (many applications apparently have different ways of handling this)
		- Discord just leaves the context menu at the same place, macOS Finder just blocks scroll input
- overlay containers blocking `OnMouseDown` (with an interceptor drawable)

Wondering if these changes are feasible:

https://github.com/Joehuu/osu-framework/compare/context-menu-container-interceptor

https://github.com/ppy/osu/compare/master...Joehuu:osu:centralise-context-menu

frenzibyte commented:
What you have looks sane on a quick whim, would need confirmation on proposed UX but take what you have into a PR and let's have it discussed there.

bdach commented:
I don't know if I'd agree. I can't very quickly say whether adding "interceptors" that "bypass" `OnMouseDown()` return values is going to be correct in every single use case of context menu containers.
